tla-pp/src/main/scala/at/forsyte/apalache/tla/pp/temporal/LoopEncoder.scala

```diff
@@ -0,0 +1,267 @@
+      variables
+        .zip(loopVariables)
+        .foldLeft(init)({ case (curInit, (varDecl, loopVarDecl)) =>
+          addLoopVarToInit(varDecl, loopVarDecl, curInit)
+        })
+
+    //  inLoop = FALSE
+    val inLoopEqlFalse =
+      builder.eql(inLoop, builder.bool(false))
+
+    /*
+        /\ inLoop = FALSE
+        /\ init
+        /\ loop_foo = foo
+        /\ loop_bar = bar
+        /\ ...
+     */
+    conjunctExToOperDecl(inLoopEqlFalse, initWithLoopVarInits, tracker)
+  }
+
+  /**
+   * Adds logic for manipulating the loop_var in next. Transforms next into
+   *
+   * newNext ==
+   *
+   * /\ oldNext
+   *
+   * /\ loop_foo' = IF (InLoop' = InLoop) THEN loop_foo ELSE foo
+   */
+  def addLoopVarToNext(varDecl: TlaVarDecl, loopVarDecl: TlaVarDecl, next: TlaOperDecl): TlaOperDecl = {
+    val loopEx = builder.declAsNameEx(loopVarDecl)
+    val loopExPrime = builder.prime(loopEx)
+
+    TlaOperDecl(
+        next.name,
+        next.formalParams,
+        builder.and(
+            /* oldNext */
+            builder.useTrustedEx(next.body),
+            /* /\ loop_foo' = IF (InLoop' = InLoop) THEN loop_foo ELSE foo */
+            builder.eql(
+                loopExPrime,
+                builder.ite(
+                    builder.eql(inLoop, inLoopPrime),
+                    loopEx,
+                    builder.declAsNameEx(varDecl),
+                ),
+            ),
+        ),
+    )(next.typeTag)
+  }
+
+  /**
+   * Modifies next to include the logic for updating InLoop and the loop variables.
+   */
+  def addLoopLogicToNext(
+      variables: Seq[TlaVarDecl],
+      loopVariables: Seq[TlaVarDecl],
+      next: TlaOperDecl): TlaOperDecl = {
+
+    /*
+     * /\ InLoop' \in {TRUE, FALSE}
+     * /\ InLoop => InLoop'
+     */
+    val inLoopUpdate =
+      builder.and(
+          builder.in(inLoopPrime, builder.booleanSet()),
+          builder.impl(inLoop, inLoopPrime),
+      )
+
+    val nextOrUnchangedWithInLoopUpdate = conjunctExToOperDecl(inLoopUpdate, next, tracker)
+
+    /*
+     * /\ loop_foo' = IF (InLoop' = InLoop) THEN loop_foo ELSE foo
+     */
+    variables.zip(loopVariables).foldLeft(nextOrUnchangedWithInLoopUpdate) { case (curNext, (varDecl, loopVarDecl)) =>
+      addLoopVarToNext(varDecl, loopVarDecl, curNext)
+    }
+  }
+
+  /**
+   * Adds a loop variable to the loopOK expression by transforming loopOK into
+   *
+   * newLoopOK ==
+   *
+   * /\ oldLoopOK
+   *
+   * /\ foo = loop_foo
+   */
+  def addVariableToLoopOK(
+      varDecl: TlaVarDecl,
+      loopVarDecl: TlaVarDecl,
+      loopOK: TlaOperDecl): TlaOperDecl = {
+
+    conjunctExToOperDecl(
+        builder.eql(builder.declAsNameEx(varDecl), builder.declAsNameEx(loopVarDecl)),
+        loopOK,
+        tracker,
+    )
+  }
+
+  /**
+   * Creates a loopOK predicate over the provided variables.
+   *
+   * loopOK ==
+   *
+   * /\ InLoop
+   *
+   * /\ foo = loop_foo
+   *
+   * /\ bar = loop_bar
+   */
+  def createLoopOKPredicate(
```

Comment:
But let me know if I misunderstood your point!

Reply:
They also use a loopOK predicate, albeit called differently:
see p166, fig 5. They `define` `looped := saved & state = loop`
This will need to be reevaluated in each step too, as far as I understand.
It is exactly the same as my predicate: `LoopOK = InLoop /\ state = loop_state`

Reply:
Maybe it's confusing because I called it a predicate? LoopOk is a 0-parameter boolean valued expression

Reply:
Oh, I think I have though about the start of the loop, and you are closing the loop here. Sorry about the confusion. It's all good.